livekit-agents/livekit/agents/pipeline/agent_output.py

```diff
@@ -187,118 +191,183 @@ async def _synthesize_task(self, handle: SynthesisHandle) -> None:
-        async for audio in handle._tts.synthesize(tts_text):
-            if first_frame:
-                first_frame = False
-                logger.debug(
-                    "received first TTS frame",
-                    extra={
-                        "speech_id": handle.speech_id,
-                        "elapsed": round(time.time() - start_time, 3),
-                        "streamed": False,
-                    },
-                )
-                read_transcript_atask = asyncio.create_task(
-                    _read_transcript_task(transcript_source, handle)
-                )
+            if not handle.tts_forwarder.closed:
+                handle.tts_forwarder.mark_text_segment_end()
+        finally:
+            if inspect.isasyncgen(transcript_source):
+                await transcript_source.aclose()
+
+    @utils.log_exceptions(logger=logger)
+    async def _str_synthesis_task(
+        self,
+        tts_text: str,
+        transcript_source: AsyncIterable[str] | str,
+        handle: SynthesisHandle,
+    ) -> None:
+        """synthesize speech from a string"""
+        read_transcript_atask: asyncio.Task | None = None
 
-            frame = audio.frame
+        start_time = time.perf_counter()
+        cancelled = False
+        audio_duration = 0.0
+        ttfb = -1.0
 
-            handle._buf_ch.send_nowait(frame)
-            if not handle.tts_forwarder.closed:
-                handle.tts_forwarder.push_audio(frame)
+        tts_stream = handle._tts.synthesize(tts_text)
+        try:
+            async for audio in tts_stream:
+                if ttfb == -1.0:
+                    ttfb = time.perf_counter() - start_time
 
-    finally:
-        if not handle.tts_forwarder.closed:
-            handle.tts_forwarder.mark_audio_segment_end()
+                    read_transcript_atask = asyncio.create_task(
+                        self._read_transcript_task(transcript_source, handle)
+                    )
 
-        if read_transcript_atask is not None:
-            await read_transcript_atask
+                audio_duration += audio.frame.duration
 
+                handle._buf_ch.send_nowait(audio.frame)
+                if not handle.tts_forwarder.closed:
+                    handle.tts_forwarder.push_audio(audio.frame)
 
-@utils.log_exceptions(logger=logger)
-async def _stream_synthesis_task(
-    tts_source: AsyncIterable[str],
-    transcript_source: AsyncIterable[str] | str,
-    handle: SynthesisHandle,
-) -> None:
-    """synthesize speech from streamed text"""
+            if not handle.tts_forwarder.closed:
+                handle.tts_forwarder.mark_audio_segment_end()
+
+            if read_transcript_atask is not None:
+                await read_transcript_atask
+        except asyncio.CancelledError:
+            cancelled = True
+        finally:
+            await tts_stream.aclose()
+
+            duration = time.perf_counter() - start_time
+
+            tts_metrics: metrics.TTSMetrics = {
```

Comment:
should we compute this in the TTS module? it would seem that it's more re-usable in future Agents